### PR TITLE
Remove discontinued kotlin-coroutines-retrofit dependency.

### LINKED ISSUE #271: Remove kotlin-coroutines-retrofit dependency
# Current state
- [`ru.gildor.coroutines:kotlin-coroutines-retrofit:1.1.0`](https://github.com/gildor/kotlin-coroutines-retrofit) is used in `EngelsystemNetworkRepository` in the `engelsystem` module. The library depends on [`org.jetbrains.kotlinx:kotlinx-coroutines-core:1.1.1`](https://github.com/Kotlin/kotlinx.coroutines/releases/tag/1.1.1) which was released 24.01.2019.
- [Development of the library is discontinued](https://github.com/gildor/kotlin-coroutines-retrofit/pull/57#pullrequestreview-331689298) because Retrofit introduced `<T : Any> Call<T>.await*()` extension functions in version [2.6.0](https://github.com/square/retrofit/blob/master/CHANGELOG.md#version-260-2019-06-05) itself. See https://github.com/square/retrofit/commit/b761518aa174c7b0512b73f2fe70e2e908f24081.
- `com.squareup.retrofit2:retrofit:2.6.4` is used as a transitive dependency of `info.metadude.kotlin.library.engelsystem:engelsystem-base:3.1.0`.

# Target state
- Use Retrofit's functions to replace `kotlin-coroutines-retrofit`.

# TODOs
- [x] Replace `kotlin-coroutines-retrofit`'s `awaitResult` with `retrofit`'s `awaitResponse` function.
- [x] Add unit tests.
- [x] Manually test if the Engelsystem feature still works. @johnjohndoe can provide access to a test backend.

### PR DESCRIPTION
**Pull request by johnjohndoe**
# Description
+ Remove discontinued `kotlin-coroutines-retrofit` dependency.
+ _Retrofit_ introduced an `<T : Any> Call<T>.awaitResponse()` extension function in version 2.6.0 itself.
  See https://github.com/square/retrofit/commit/b761518aa174c7b0512b73f2fe70e2e908f24081.

# Successfully tested on
with `ccc36c3` flavor, `debug` build
- :heavy_check_mark: Google Pixel 2, Android 10.

# Review
- To manually test if the _Engelsystem_ feature still works reach out to @johnjohndoe who can provide access to a test backend.

---

Resolves #271.

### CODE REVIEW

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        runBlocking {
+            callBlock.invoke(call)
+        }
+    }
+
+    /**
+     * Only use for success test cases because [Callback.onResponse] is always invoked!
+     * The [response] property must be set in each test.
+     */
+    private class SuccessCall : Call<List<Shift>> {
```
Comment (cketti) [inline engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt:154]: If you split out a `Response.toShiftsResult()` extension method as suggested above, you don't have to create a `SuccessCall` instance just to test the transformation code.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    @Test
+    fun `awaitShiftsResult returns MissingResponseSuccessBody exception when call responds with null body`() {
+        withSuccessResponse(Response.success(null)) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isEqualTo(ShiftsResult.Exception.MissingResponseSuccessBody)
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns success when call responds with HTTP 200`() {
```
Comment (cketti) [inline engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt:66]: If you write it like this:

```kotlin
@Test
fun `awaitShiftsResult returns success when call responds with HTTP 200`() = runBlocking {
    val call = performHttpRequest(HttpURLConnection.HTTP_OK, VALID_SHIFTS_JSON)

    val shiftsResult = call.awaitShiftsResult()

    assertThat(shiftsResult).isEqualTo(ShiftsResult.Success(ONE_ITEM_SHIFTS))
}
```

the arrange, act, and assert phase can be visually separated quite nicely.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        withResponse(HttpURLConnection.HTTP_OK) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isEqualTo(ShiftsResult.Success(ONE_ITEM_SHIFTS))
```
Comment (cketti) [inline engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt:69]: If you assert against `ONE_ITEM_SHIFTS` I'd pass `VALID_SHIFTS_JSON` to `withResponse`. I try to only use default arguments for parameters that are not relevant to the test. 
By only reading the test and not the implementation of `withResponse()` the reader has no idea why the result should be `ONE_ITEM_SHIFTS`.

Side note: Maybe the names of the constants should be changed so it's clear it's the same data in two different representations.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+import org.junit.Test
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+import java.io.EOFException
+import java.net.HttpURLConnection
+
+class CallExtensionsTest {
```
Comment (cketti) [inline engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt:22]: A test with invalid JSON data (both non-JSON data and valid JSON data that doesn't match the "schema") would be nice.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+    companion object {
+        private const val VALID_SHIFTS_JSON = "[{\"Comment\":\"This is a very secret comment.\",\"Name\":\"Kirmes\",\"RID\":12,\"SID\":579,\"TID\":10,\"UID\":32,\"created_at_timestamp\":1565979486,\"created_by_user_id\":32,\"description\":\"Kirmes are fun.\",\"edited_at_timestamp\":1567367095,\"edited_by_user_id\":32,\"freeloaded\":0,\"id\":37,\"name\":\"Collect stickers\",\"shifttype_id\":6,\"title\":\"Tag 1: Decorate fridge\"}]"
```
Comment (cketti) [inline engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt:25]: Probably looks nicer when using a raw string literal and line breaks.

File: engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt [hunk 6]
```diff
@@ -4,8 +4,28 @@ import info.metadude.kotlin.library.engelsystem.models.Shift
 
 sealed class ShiftsResult {
 
-    data class Success(val shifts: List<Shift>) : ShiftsResult()
-    data class Error(val httpStatusCode: Int, val exceptionMessage: String) : ShiftsResult()
-    data class Exception(val throwable: Throwable) : ShiftsResult()
+    data class Success(val shifts: List<Shift>) : ShiftsResult() {
+
+        override fun toString() = "Success(shifts=$shifts)"
```
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt:9]: In the case of a test failure the output differs:

# Without `toString` overwrites:

``` kotlin
expected: info.metadude.android.eventfahrplan.engelsystem.models.ShiftsResult$Exception@2f490758
but was : info.metadude.android.eventfahrplan.engelsystem.models.ShiftsResult$Exception@1188e820
```

# With `toString` overwrites:

``` kotlin
expected: Exception(throwable=com.squareup.moshi.JsonDataException)
but was : Exception(throwable=com.squareup.moshi.JsonDataException: Expected BEGIN_ARRAY but was BEGIN_OBJECT at path $)
```

File: engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensions.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import info.metadude.android.eventfahrplan.engelsystem.models.ShiftsResult
+import info.metadude.kotlin.library.engelsystem.models.Shift
+import retrofit2.Call
+import retrofit2.awaitResponse
+
+/**
+ * Suspend extension that allows to suspend a shifts [Call] inside a coroutine.
+ * Returns a sealed [ShiftsResult] object.
+ */
+internal suspend fun Call<List<Shift>>.awaitShiftsResult(): ShiftsResult {
```
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensions.kt]: :heavy_check_mark: Split in 0834d19ca714e71a4ebb347a4ccbe518625b3bfe.

File: engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt [hunk 8]
```diff
@@ -4,8 +4,28 @@ import info.metadude.kotlin.library.engelsystem.models.Shift
...
+
+    }
+
+    data class Error(val httpStatusCode: Int, val exceptionMessage: String) : ShiftsResult() {
+
+        override fun toString() = "Error(httpStatusCode=$httpStatusCode, exceptionMessage='$exceptionMessage')"
```
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt:15]: :x: Kept as is. See above.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt (hunk 1 above)
Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt]: :heavy_check_mark: Split and refactored new tests in 0834d19ca714e71a4ebb347a4ccbe518625b3bfe.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt (hunk 3 above)
Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt]: :heavy_check_mark: I thought about this, too. Changed in 0834d19ca714e71a4ebb347a4ccbe518625b3bfe.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt (hunk 4 above)
Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt]: :heavy_check_mark: Good idea. Done in 0834d19ca714e71a4ebb347a4ccbe518625b3bfe.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt (hunk 5 above)
Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt]: :heavy_check_mark: True. More readable in 0834d19ca714e71a4ebb347a4ccbe518625b3bfe.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt (hunk 2 above)
Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt]: Is 0834d19ca714e71a4ebb347a4ccbe518625b3bfe doing what you had in mind?

File: engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (cketti) [reply engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt:9]: If you look closer you will notice that I didn't comment on the `toString()` method of `Exception` (which is not a data class) :smile:

**Comment by johnjohndoe** [comment]
Here is the current test coverage:

image

File: engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt]: Now I see it! :heavy_check_mark: Removed in b049d870964e0bfeb1afc099f8dd0de0c4fc926b. Thank you.

File: engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (johnjohndoe) [reply engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/models/ShiftsResult.kt]: :heavy_check_mark: Removed in b049d870964e0bfeb1afc099f8dd0de0c4fc926b.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    fun `awaitShiftsResult returns success when call responds with HTTP 200`() {
+        withResponse(HttpURLConnection.HTTP_OK, VALID_ONE_ITEM_SHIFTS_JSON) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isEqualTo(ShiftsResult.Success(ONE_ITEM_SHIFTS))
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns error when call responds with HTTP 300`() {
+        withResponse(HttpURLConnection.HTTP_MULT_CHOICE, VALID_ONE_ITEM_SHIFTS_JSON) { call ->
```
Comment (cketti) [inline engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt:134]: `VALID_ONE_ITEM_SHIFTS_JSON` doesn't seem relevant for the error cases. It's also unlikely the error bodies will contain a valid JSON response in practice. In this case a default value for the `shiftsJson` parameter would be good.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    @Test
+    fun `Converts a success response with null body into an ShiftsResult exception`() {
+        val response: Response<List<Shift>> = Response.success(null)
+        val shiftsResult = response.toShiftsResult()
+        assertThat(shiftsResult).isEqualTo(ShiftsResult.Exception.MissingResponseSuccessBody)
+    }
+
+    @Test
+    fun `Converts a success response with empty body into an ShiftsResult success with an empty list`() {
+        val shiftsResult = Response.success(emptyList<Shift>()).toShiftsResult()
```
Comment (cketti) [inline engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt:88]: I find it enormously helpful to separate tests into arrange, act, and assert sections. That way it's clear what you're testing exactly, i.e. which method. With a clear separation of inputs and expected results.

Here you're mixing the first two phases and I find it requires much more effort to read the test.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt (hunk 2 above)
Comment (cketti) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt]: Not really. When you're doing `someMethod(args) { someObject -> someMoreCode(); someAssertion(); }` I'm not really sure if `someMethod()` just executes the lambda or if it also does some important work afterwards. 
With the proposed code I find it's clear what `performHttpRequest()` does. I'm not tempted to look at the implementation to check out the details when reading an individual test. It also allows the nice separation into arrange, act, and assert sections without any nesting.

Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt]: I actually started with the proposed syntax and factored out the repetitive `runBlocking` later. I however also like the proposal. I changed the test in d616cee9a6bd0f59e474a9a4e7c92efc45792b7a.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt (hunk 9 above)
Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt]: :heavy_check_mark: Done in d616cee9a6bd0f59e474a9a4e7c92efc45792b7a.

File: engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt (hunk 10 above)
Comment (johnjohndoe) [reply engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt]: :heavy_check_mark: Separated in d616cee9a6bd0f59e474a9a4e7c92efc45792b7a.

**Comment by johnjohndoe** [comment]
# Successfully retested on
with `ccc36c3` flavor, `release` build
- :heavy_check_mark: Google Pixel 2, Android 10.
- :heavy_check_mark: HTC Nexus 9, Android 7.1.1.

**Comment by cketti** [comment]
:+1:

**Pull request merged**